In [6]:
import pandas as pd

In [7]:
data = pd.read_csv("/content/exams.csv")

In [8]:
data

gender race/ethnicity parental level of education         lunch  \
0      male        group A                 high school      standard   
1    female        group D            some high school  free/reduced   
2      male        group E                some college  free/reduced   
3      male        group B                 high school      standard   
4      male        group E          associate's degree      standard   
..      ...            ...                         ...           ...   
995    male        group C                 high school      standard   
996    male        group D          associate's degree  free/reduced   
997  female        group C            some high school  free/reduced   
998  female        group C                some college      standard   
999    male        group A                some college      standard   

    test preparation course  math score  reading score  writing score  
0                 completed        67.0           67.0           63.0  
1                      none        40.0           59.0           55.0  
2                      none        59.0           60.0           50.0  
3                      none        77.0           78.0           68.0  
4                 completed        78.0           73.0           68.0  
..                      ...         ...            ...            ...  
995                    none        73.0           70.0           65.0  
996               completed        85.0           91.0           92.0  
997                    none        32.0           35.0           41.0  
998                    none        73.0           74.0           82.0  
999               completed        65.0           60.0           62.0  

[1000 rows x 8 columns]

In [9]:
data.dtypes

gender                          object
race/ethnicity                  object
parental level of education     object
lunch                           object
test preparation course         object
math score                     float64
reading score                  float64
writing score                  float64
dtype: object

In [10]:
data.head()

gender race/ethnicity parental level of education         lunch  \
0    male        group A                 high school      standard   
1  female        group D            some high school  free/reduced   
2    male        group E                some college  free/reduced   
3    male        group B                 high school      standard   
4    male        group E          associate's degree      standard   

  test preparation course  math score  reading score  writing score  
0               completed        67.0           67.0           63.0  
1                    none        40.0           59.0           55.0  
2                    none        59.0           60.0           50.0  
3                    none        77.0           78.0           68.0  
4               completed        78.0           73.0           68.0

In [11]:
data.isnull()

gender  race/ethnicity  parental level of education  lunch  \
0     False           False                        False  False   
1     False           False                        False  False   
2     False           False                        False  False   
3     False           False                        False  False   
4     False           False                        False  False   
..      ...             ...                          ...    ...   
995   False           False                        False  False   
996   False           False                        False  False   
997   False           False                        False  False   
998   False           False                        False  False   
999   False           False                        False  False   

     test preparation course  math score  reading score  writing score  
0                      False       False          False          False  
1                      False       False          False          False  
2                      False       False          False          False  
3                      False       False          False          False  
4                      False       False          False          False  
..                       ...         ...            ...            ...  
995                    False       False          False          False  
996                    False       False          False          False  
997                    False       False          False          False  
998                    False       False          False          False  
999                    False       False          False          False  

[1000 rows x 8 columns]

In [12]:
data['math score'].fillna(data['math score'].mean(), inplace=True)

In [13]:
data['reading score'].fillna(data['reading score'].mean(), inplace=True)

In [14]:
data['writing score'].fillna(data['writing score'].mean(), inplace=True)

In [15]:
data['gender'].fillna(data['gender'].mode()[0], inplace=True)

In [16]:
data['race/ethnicity'].fillna(data['race/ethnicity'].mode()[0], inplace=True)


In [17]:
data['parental level of education'].fillna(data['parental level of education'].mode()[0], inplace=True)


In [18]:
data['lunch'].fillna(data['lunch'].mode()[0], inplace=True)

In [19]:
data['test preparation course'].fillna(data['test preparation course'].mode()[0], inplace=True)

In [20]:
from sklearn.preprocessing import LabelEncoder

le=LabelEncoder()

In [21]:
data['gender'] = le.fit_transform(data['gender'])
data['race/ethnicity'] = le.fit_transform(data['race/ethnicity'])
data['parental level of education'] = le.fit_transform(data['parental level of education'])
data['lunch'] = le.fit_transform(data['lunch'])
data['test preparation course'] = le.fit_transform(data['test preparation course'])


In [23]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import  train_test_split
from sklearn.metrics import accuracy_score

# Split features and target
X = data.drop(columns=['writing score'])
y = data['writing score']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

# Define the Decision Tree classifier with desired parameters
classifier = DecisionTreeClassifier(max_depth=5, criterion='gini')

# Train the classifier
classifier.fit(X_train, y_train)

# Predict on the test set
y_pred = classifier.predict(X_test)

In [24]:
y_test

521    82.0
737    81.0
740    51.0
660    97.0
411    72.0
       ... 
109    33.0
430    85.0
77     81.0
84     39.0
286    84.0
Name: writing score, Length: 250, dtype: float64

In [25]:
y_pred

array([ 77.,  80.,  47.,  84.,  75.,  71., 100.,  80.,  80.,  80.,  63.,
        63.,  53.,  71.,  40.,  40.,  80.,  80.,  80.,  80.,  56.,  75.,
        66.,  84.,  65.,  84.,  80.,  77.,  66.,  63.,  80.,  71.,  48.,
        80.,  71.,  63.,  65.,  80.,  65.,  48.,  61., 100.,  47.,  62.,
        40.,  80.,  63.,  71.,  63.,  66.,  63.,  47.,  47.,  66.,  56.,
        80.,  63.,  80.,  65.,  80.,  48.,  53.,  89.,  63.,  95.,  84.,
        88.,  53.,  63.,  61.,  40.,  80.,  82.,  63.,  71.,  66., 100.,
        80.,  71.,  63.,  62.,  62.,  80.,  48.,  89.,  47.,  71.,  71.,
        62.,  66.,  80.,  53.,  71.,  61.,  80.,  71.,  77.,  48.,  80.,
        95.,  89.,  48.,  48.,  80.,  80.,  47.,  77.,  80.,  89.,  71.,
        61.,  66.,  71., 100.,  82.,  71.,  80.,  47.,  77.,  61.,  40.,
        80.,  56.,  66.,  88.,  48.,  80.,  71.,  82.,  80.,  89.,  66.,
        71.,  71.,  63.,  61.,  71.,  56.,  47.,  66.,  84.,  71.,  89.,
       100.,  40.,  66.,  47.,  39.,  75.,  82.,  6

In [26]:
y

0      63.0
1      55.0
2      50.0
3      68.0
4      68.0
       ... 
995    65.0
996    92.0
997    41.0
998    82.0
999    62.0
Name: writing score, Length: 1000, dtype: float64

In [31]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

model.fit(X_train, y_train)


LinearRegression()

In [32]:
from sklearn.metrics import mean_squared_error, r2_score

y_pred = model.predict(X_test)
rmse = mean_squared_error(y_test, y_pred, squared=False)
r2 = r2_score(y_test, y_pred)

print("Model Evaluation:")
print("RMSE:", rmse)
print("R2 Score:", r2)


Model Evaluation:
RMSE: 3.988364199921894
R2 Score: 0.9354052811634433


These results suggest that the Linear Regression model is performing quite well, with a low RMSE and a high R2 score. It means that the model is able to explain a significant portion (93.54%) of the variance in the Writing Score based on the features provided.



